src/editor/ui_menus/scenario_tool_infrastructure_options_menu.cc

```diff
@@ -0,0 +1,407 @@
+			i->icon_clicked.connect(
+			   boost::bind(&ScenarioToolInfrastructureOptionsMenu::toggle_selected, this, i,
+			               Widelands::MapObjectType::FLAG, _1));
+			tab->add("flag", icon, i, _("Flag"));
+		}
+		{
+			UI::IconGrid* i =
+			   new UI::IconGrid(tab, 0, 0, kIconGridCellSize, kIconGridCellSize, kIconGridColumns);
+			for (Widelands::DescriptionIndex di : tribe.immovables()) {
+				const Widelands::ImmovableDescr* d = tribe.get_immovable_descr(di);
+				i->add(d->name(), d->icon(), reinterpret_cast<void*>(di), d->descname());
+			}
+			i->icon_clicked.connect(
+			   boost::bind(&ScenarioToolInfrastructureOptionsMenu::toggle_selected, this, i,
+			               Widelands::MapObjectType::IMMOVABLE, _1));
+			tab->add("immovables", g_gr->images().get("images/wui/menus/toggle_immovables.png"), i,
+			         _("Immovables"));
+		}
+		tabs_.add("player_" + std::to_string(static_cast<unsigned>(p)),
+		          map.get_scenario_player_name(p), tab);
+	}
+	tabs_.sigclicked.connect(boost::bind(&ScenarioToolInfrastructureOptionsMenu::select_tab, this));
+
+	force_.set_state(tool_.get_force());
+	force_.changed.connect([this]() {
+		tool_.set_force(force_.get_state());
+		select_correct_tool();
+	});
+	auto_infra_cur_.sigclicked.connect([this]() {
+		make_auto_infra(false);
+		select_correct_tool();
+	});
+	auto_infra_all_.sigclicked.connect([this]() {
+		make_auto_infra(true);
+		select_correct_tool();
+	});
+
+	buttonsbox_.add(&auto_infra_cur_, UI::Box::Resizing::kExpandBoth);
+	buttonsbox_.add(&auto_infra_all_, UI::Box::Resizing::kExpandBoth);
+	box_.add(&tabs_, UI::Box::Resizing::kExpandBoth);
+	box_.add(&force_, UI::Box::Resizing::kFullSize);
+	box_.add(&buttonsbox_, UI::Box::Resizing::kFullSize);
+	box_.add(&selected_items_, UI::Box::Resizing::kFullSize);
+	tabs_.activate(tool_.get_player() - 1);
+	select_tab();
+	set_center_panel(&box_);
+	layout();
+	if (get_usedefaultpos()) {
+		center_to_parent();
+	}
+}
+
+void ScenarioToolInfrastructureOptionsMenu::make_auto_infra(bool all) {
+	std::list<std::string> errors;
+	Widelands::EditorGameBase& egbase = eia().egbase();
+	auto is_hq = [&egbase](Widelands::DescriptionIndex di) {
+		const Widelands::BuildingDescr& b = *egbase.tribes().get_building_descr(di);
+		return b.type() == Widelands::MapObjectType::WAREHOUSE &&
+		       b.get_size() == Widelands::BaseImmovable::Size::BIG && !b.get_isport() &&
+		       !b.get_ismine() && !b.is_destructible() && !b.can_be_dismantled() &&
+		       !b.is_buildable() && !b.is_enhanced() && !b.needs_seafaring() &&
+		       !b.needs_waterways() && b.get_built_over_immovable() == Widelands::INVALID_INDEX;
+	};
+	std::set<unsigned> for_whom;
+	if (all) {
+		for (unsigned p = egbase.map().get_nrplayers(); p; --p) {
+			for_whom.insert(p);
+		}
+	} else {
+		for_whom.insert(tabs_.active() + 1);
+	}
+	for (unsigned p : for_whom) {
+		Widelands::Player& player = *egbase.get_player(p);
+		Widelands::DescriptionIndex index =
```

Comment:
You can replace this hack by detecting the warehouse building:

`building.descr().type() == MapObjectType::WAREHOUSE && !building.is_buildable() && !building.is_enhanced()`

Reply:
For the existing tribes it would be enough to use "tribename_headquarters". But future tribes might have some really strange buildings. E.g. they *could* have a non-buildable, non-enhanced warehouse which might e.g. have `mine` size or is only enabled on seafaring maps for whatever reason. So I'd rather use a bunch of sanity checks so this works with all imaginable custom tribes as well.

Reply:
Add !is_mine() and !is_port() to the detection then.

Alternatively, add a special building "headquarters" to TribeDescr, then we know that we have one and can get rid of all the checking and guessing.

Reply:
ismine and isport checks are already part of the checksuite in is_hq() ;)

Reply:
You're right. I still don't like depending on building names - it would be easer tp require that all tribes have a special building "headquarters". We actually used to have that code but removed it, because it was unused.

If you want, we can just add a "TODO" comment and I'll fix this up in my tribes-filtering branch to avoid merge conflicts.